web/src/test/java/org/springframework/security/web/server/csrf/CsrfWebFilterTests.java

```diff
@@ -227,6 +230,31 @@ public void doFilterWhenTokenIsNullThenNoNullPointer() {
 				.isForbidden();
 	}
 
+	@Test
+	public void filterWhenSubscribingCsrfTokenMultipleTimesThenGenerateOnlyOnce() {
+		this.csrfFilter.setCsrfTokenRepository(this.repository);
+		given(this.repository.loadToken(any())).willReturn(Mono.empty());
+		AtomicInteger callGenerateToken = new AtomicInteger(0);
+		given(this.repository.generateToken(any())).willReturn(Mono.defer(() -> {
+			callGenerateToken.incrementAndGet();
+			return Mono.just(this.token);
+		}));
+		AtomicInteger callSaveToken = new AtomicInteger(0);
+		given(this.repository.saveToken(any(), any())).willReturn(Mono.defer(() -> {
+			callSaveToken.incrementAndGet();
+			return Mono.empty();
+		}));
+		AtomicReference<Mono<CsrfToken>> tokenFromExchange = new AtomicReference<>();
+		given(this.chain.filter(any())).willReturn(
+				Mono.fromRunnable(() -> tokenFromExchange.set(this.get.getAttribute(CsrfToken.class.getName()))));
+		this.csrfFilter.filter(this.get, this.chain).block();
+		tokenFromExchange.get().block();
```

Comment:
Instead of incrementing counters, would it be better to block multiple times and confirm that you receive the same value? For example:

```java
CsrfToken firstToken = tokenFromExchange.get().block();
CsrfToken secondToken = tokenFromExchange.get().block();
assertThat(firstToken).isEqualTo(secondToken);
```

This might simplify some of the instrumentation you are doing in the test. Also, since calling `saveToken` multiple times is not necessarily a problem for this test, removing it may make the test more resilient.

Reply:
Thank you for your review and suggestions.
I'm working now.